pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,309 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Lie gradient optimizers"""
+import numpy as np
+from scipy.sparse.linalg import expm
+import pennylane as qml
+from pennylane.tape import JacobianTape
+from pennylane.transforms import batch_transform
+
+
+@qml.qfunc_transform
+def append_time_evolution(tape, lie_gradient, t, exact=False):
+    r"""Append an approximate time evolution to an existing circuit.
+    If `exact` is `False`, we Trotterize the Hamiltonian and with a single step
+
+    .. math:
+
+        U' = \prod_i \exp{-it O_i}
+
+    Then this unitary is appended to the current circuit.
+
+    If `exact` is `True`, we calculate the exact time evolution for the Lie gradient by way of the
+    matrix exponential.
+
+    .. math:
+
+        U' = \exp{-it \text{grad} f(U)}
+
+    and append this unitary.
+
+    Args:
+        tape: qml.QuantumTape of circuit
+        lie_gradient: qml.Hamiltonian object
+        t: time evolution size
+
+    """
+    for obj in tape.operations:
+        qml.apply(obj)
+    if exact:
+        qml.QubitUnitary(
+            expm(t * qml.utils.sparse_hamiltonian(lie_gradient)).toarray(),
+            wires=range(max(lie_gradient.wires) + 1),
+        )
+    else:
+        qml.templates.ApproxTimeEvolution(lie_gradient, t, 1)
+
+    for obj in tape.measurements:
+        qml.apply(obj)
+
+
+@batch_transform
+def algebra_commutator(tape, observables, lie_algebra_basis_names, nqubits):
+    """
+    Calculate the Lie gradient with the parameter shift rule (see `get_omegas`).
+
+    Args:
+        tape (qml.QuantumTape):
+        observables (list):
+        lie_algebra_basis_names (list):
+        nqubits (int):
+
+    Returns:
+        list of `qml.QuantumTape`'s and None
+    """
+    tapes = []
+    for obs in observables:
+        # create a list of tapes for the plus and minus shifted circuits
+        tapes_plus = [JacobianTape(p + "_p") for p in lie_algebra_basis_names]
+        tapes_min = [JacobianTape(p + "_m") for p in lie_algebra_basis_names]
+
+        # loop through all operations on the input tape
+        for op in tape.operations:
+            for t in tapes_plus + tapes_min:
+                with t:
+                    qml.apply(op)
+        for i, t in enumerate(tapes_plus):
+            with t:
+                qml.PauliRot(
+                    np.pi / 2,
+                    lie_algebra_basis_names[i],
+                    wires=list(range(nqubits)),
+                )
+                for o in obs:
+                    qml.expval(o)
+        for i, t in enumerate(tapes_min):
+            with t:
+                qml.PauliRot(
+                    -np.pi / 2,
+                    lie_algebra_basis_names[i],
+                    wires=list(range(nqubits)),
+                )
+                for o in obs:
+                    qml.expval(o)
+        tapes.append((tapes_plus, tapes_min))
+    return tapes, None
```

Comment:
Maybe. I made an attempt to simply return a single flattened list and divide them up after execution, but ran into problems. I will make another attempt to make this nicer.

Reply:
@therooler now that I've read through the full file, I don't think you need the `batch_transform` decorator here, since you are using this function as-is on a tape. All the `batch_transform` decorator does is allow you to also use this function on a _qnode_, which doesn't seem to be occurring :)